src/lib/scheduler/abstract_task.cpp

```diff
@@ -36,11 +39,12 @@ std::string AbstractTask::description() const {
 void AbstractTask::set_id(TaskID id) { _id = id; }
 
 void AbstractTask::set_as_predecessor_of(const std::shared_ptr<AbstractTask>& successor) {
-  Assert((!_is_scheduled), "Possible race: Don't set dependencies after the Task was scheduled");
-
-  successor->_pending_predecessors++;
-  _successors.emplace_back(successor);
-  successor->_predecessors.emplace_back(shared_from_this());
+  if (std::find(_successors.cbegin(), _successors.cend(), successor) == _successors.cend()) {
+    _successors.emplace_back(successor);
+    successor->_predecessors.emplace_back(shared_from_this());
+    auto lock = std::unique_lock<std::mutex>(_done_condition_variable_mutex);
+    if (!is_done()) successor->_pending_predecessors++;
+  }
```

Comment:
Could you explain these changes, please?

Reply:
When the ExpressionEvaluator calls `make_tasks_from_operator`, it receives the full set of tasks. However, some of them may already have executed. Since finished tasks will never call `_on_predecessor_done`, we would wait indefinitely for `_pending_predecessors` to reach 0.

Reply:
That makes sense. Why do we now need to check if the task is already part of `_successors` and why didn't we have to do it before?

Reply:
Because tasks now get reused.

Compare plan from https://github.com/hyrise/hyrise/issues/2014#issuecomment-800107659 (SQLiteTestRunner391):

1. Root query gets constructed: Validate calls `set_as_predecessor_of` at AggregateHash and adds it to its successors.
2. Root query gets executed: The TableScan calls the ExpressionEvaluator, which then calls `make_tasks_from_operators`.
3. Uncorrelated query gets constructed: The same Validate task (from 1.) calls `set_as_predecessor_of` at AggregateHash and adds it to its successors. Again!

I found numerous SQLiteTestRunner queries, which take the following path:

```cpp
if (std::find(_successors.cbegin(), _successors.cend(), successor) != _successors.cend()) return;
``` 

For example: 94, 100, 122, 138-139, 169-170, 173-176, 188-193, 195-199, 205, 207, 212, 228, 273, 389-391, 398

Interestingly, all tests still pass after commenting it out. But, I would keep it anyway.